### ISSUE TITLE
Improve documentation to provide more details to work in a case of disaster recovery

### ISSUE THREAD

**Issue by weslleycamilo** [bitnami-labs/sealed-secrets#357]
Hi guys,

I've been reading the docs and I found some good details which helped me to test it.

What about to **include more details about a case of disaster of a K8s** cluster :thinking: 

I've been testing the sealed secrets and I've considering use its key renewal but faced some doubts, some of them are:

- need to re-encrypt everything and I don't know a good way to do it to many secrets once it looks like I'll have to automate it to find all the secrets on kubernetes and re-encrypt it right ? or there is a flag ?

- If I would like to create my own key and not renew it can I do that ? Would I have this flexibility ? I considered it after not know what to do in case I lose my whole cluster and don't have all the keys which was lost in the broken cluster. So if I have my own key it won't be a problem to encrypt, decrypt and launch a new cluster which can decrypt my sealedsecrets to the cluster from github whith Fluxcd.

So my building/testing scenario is I've all the sealed secrets in the Github and in case of launch/migration to new cluster I would like to just deploy FluxCD and Sealed secrets and get everything on my k8s manifests deployed ASAP and I'm not sure how to do it which sealed renew once I might lose it in the broken cluster, or have to manually re-encrypt everything :(  It looks easier to work on if I can create and use my own keys and use in all my cluster and when I would like to re-encrypt it I can do but in a planned way and not get surprised by  a disaster recovery.

**Comment by mkmik** [comment]
> need to re-encrypt everything and I don't know a good way to do it to many secrets once it looks like I'll have to automate it to find all the secrets on kubernetes and re-encrypt it right ? or there is a flag ?

The key renewal feature doesn't require re-encrypting everything.

Could you please help me figure out what part of the documentation makes you think that the automatic renewal feature requires you to re-encrypt everything?

The backup process described in https://github.com/bitnami-labs/sealed-secrets#how-can-i-do-a-backup-of-my-sealedsecrets is fully compatible with automatic key renewal; obviously you have to backup periodically.

**Comment by weslleycamilo** [comment]
One of the places I got its doubt was 
- https://github.com/bitnami-labs/sealed-secrets#re-encryption-advanced which say "Before you can get rid of some old sealing keys you need to re-encrypt your SealedSecrets with the latest private key)."
- https://github.com/bitnami-labs/sealed-secrets#sealing-key-renewal which say "Sealed secrets are not automatically rotated and old keys are not deleted when new keys are generated. Old sealed secrets resources can be still decrypted (that's because old sealing keys are not deleted)."

So I understood that if I use key rotation and there is a new key I must re-encrypt everything to the last key or I'll have multiple secrets with multiple key rotation If I rotate it each 30 days for example.

It looks like I would get lost with those many keys and secrets. did you get the point ? in case of disaster recovery it would be painful. 

Possibilities I see I could do from the doc and talking with you:
- Not update the keys to often and when I do I should re-encrypt everything so I make sure to not get lost with the keys/secrets.
- This doc below helped me to understand that I can share the keys between k8s clusters, for that I've to backup all keys and store them in a safe place right ? It is a good option 

https://github.com/bitnami-labs/sealed-secrets#manual-key-management-advanced


Now I understand that there is this possibilities to work in a disaster recovery. 

My suggestion than would be to include in the section of https://github.com/bitnami-labs/sealed-secrets#sealing-key-renewal to include more details like:
> Do not delete the keys before backup it or re-encrypt all the secrets with the new keys or you must lose the secret.


I'm also going to open a PR for you all with a test I've been doing here and if it works and looks good to be inserted than it might help some folks.

Thank you @mkmik

**Comment by weslleycamilo** [comment]
Related to https://github.com/bitnami-labs/sealed-secrets/pull/358

**Comment by mkmik** [comment]
> Before you can get rid of some old sealing keys you need to re-encrypt your SealedSecrets with the latest private key).

This sentence can be also written like:

"You *only* have to re-encrypt your SealedSecrets *if* you delete some old sealing keys"

Normally you should not delete old sealing keys.
The [Common misconceptions about key renewal](https://github.com/bitnami-labs/sealed-secrets#common-misconceptions-about-key-renewal) section offers a detailed explanation on why deleting old keys is not the proposed workflow.

Of course, people, being people, have different opinions and traditions; so we do allow people to do whatever they want. That's why there are sections in the doc (hopefully all labeled as *advanced* ) explaining how to deal with an unsupported workflow. You can disable automatic renewal, you can manually create/update/delete encryption keys and we offer some primitives to deal with the consequences.

Perhaps we should make it more clear and split the "supported workflow" from the "creative workflows" in our docs.

> So I understood that if I use key rotation and there is a new key I must re-encrypt everything to the last key or I'll have multiple secrets with multiple key rotation If I rotate it each 30 days for example.

Yes, if you use key rotation there is a new key every month or so.

After 10 years your k8s cluster will have 120 k8s secret resources, taking up ~840Kb of space uncompressed (or ~12K gzipped).

Compared to the amount of other manifests your cluster is likely going to contain, that's negligible. The k8s apiserver and your backup system can handle that. (I just looked at the number of service account secrets in a freshly installed GKE cluster and it's already ~40).

> It looks like I would get lost with those many keys and secrets. did you get the point ? in case of disaster recovery it would be painful.

I'm not sure I get the point.

From your point of view, as an admin, it doesn't really matter if it's one or 120 secrets.

Just run:

```
kubectl get secret -n kube-system -l sealedsecrets.bitnami.com/sealed-secrets-key -o yaml
```

and you get one single big file that contains the *sealed secret state*.
That file can be restored in one go using `kubectl apply` on a fresh cluster.
That file can be passed to `kubeseal --recovery-unseal --recovery-private-key all-keys.yaml` and it will work.

All you need to do is to make sure that you back it up.
This state will grow every 30d. So you need to make sure your back up has a recent snapshot of that state.

If you take a backup today, create a sealed secrets today, deploy your stuff on production and let it run for 6 months when disaster strikes, you'll restore it and it will work!

If you take a backup today, wait 2 months, create a sealed secrets, deploy your stuff on production and after 4 months disaster strikes, you won't be able to restore.

If you take a backup frequently enough (e.g. daily, weekly; the more frequent the better), then practically speaking, no matter when you create your sealed secrets, whenever disaster strikes you'll be able to restore.

All it takes is to back up that little amount of data. No need to do anything more.

So, basically, the only workflow that key renewal is incompatible with is if you want to do "one backup for life" and keep creating and updating sealed secrets. (That would be a bad workflow. I'm happy to elaborate on that, if needed).

If your backup automation system fails, or if you don't even have backup automation, there is one thing you have to fear not: what doesn't change can be restored. If you let some software run on production and you moved on, focusing on your next problem, you know you're covered. Only when you rotate some of your *user secrets* or you create more secrets, only then you need double-check that your backup policies are met. For example you could put that in your release checklist.

**Comment by weslleycamilo** [comment]
I may misunderstood how sealed secrets works so I'm going to read more about it and test more.

From now I've created my own key and tested it because I'm afraid of during the Disaster recovery I don't have all the keys and be unable to create all the secrets to the k8s cluster. But it is something I might will be able to use after I automate the backup of the keys and re-encrypt all sealed secrets.

As I see if I keep all the secrets updated, I mean encrypted with the latest key It would be easier once I would keep least of keys to recovery in a new k8s cluster.


I believe we can close this issue and the PR I've opened than.
